### PR TITLE
[v0.20][BUILD] - Optimize includes to reduce compile time

### PR DESCRIPTION
**Pull request by vosk**
**Only preprocessor changes in this changeset**
The original goal of this changeset is to eliminate the cycling include between Extension and ExtensionContainer, but I took the time to eliminate the include avalanche mainly in 
Link
ExtensionContainer
App/Document
Gui/Document
ViewProviderLink
ViewProviderPythonFeature
Tree

The result is measureable improvement in build time, at the expense of a few compulsory includes in a few files
The only risk is that App/Application, Base/Interpreter, Base/Reader+Writer  are not cascaded everywhere, so I tried to find all problematic files, but I dont have access to a windows machine to be 100% . I am relying on travis for this

Remove unused includes from
App/Document.h
App/Extesion.h
App/TreeView.h
App/Link.h
App/ViewProviderLink.h
App/FeaturePython.h
App/ExtensionContainer.h
Gui/Document.h
Gui/ViewProviderLink.h
Gui/ViewProviderDocumentObject.h
Base/Handle.h

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [x] Branch rebased on latest master `git pull --rebase upstream master`
- [x] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

### CODE REVIEW

**Review by chennes — APPROVED** [review]
I tested on Windows 10 and it does build successfully for me. Thanks for taking this on, any improvement in compile time is much appreciated!

**Comment by vosk** [comment]
@chennes Thank you for taking the time! Is it any faster on windows? I am hoping for ~10% reduction

**Comment by vosk** [comment]
Well, it seems my advertised speedup was not met by travis, but I do believe it is a step in the right direction

**Comment by chennes** [comment]
I tested it on a few Win10 machines this afternoon, building from scratch each time. The timings are:
HEAD: 34:00, PR4302: 33:04 (Core i5 7500 @ 3.4 GHz)
HEAD: 45:19, PR4302: 44:38 (Core i5 7400 @ 3.2 GHz)
HEAD: 45:26, PR4304: 44:14 (Core i5 7400 @ 3.2 GHz)

So faster in each case, but not a LOT faster. Are you faster on Linux?

**Comment by vosk** [comment]
On a Ryzen 3600 running Linux Mint 20 : `time make -j12` 
Before:
```
real	14m50,329s
user	147m15,533s
sys	15m42,048s
```
After:
```
real	14m32,402s
user	144m35,474s
sys	15m20,335s
```
So really, ~2%, same thing as you. I have more changes to do though :smile:

**Comment by vosk** [comment]
I've done some more cleaning up, about 5-6% faster than master now, Windows looks particularly happy.  Some conclusions could be drawn that are somewhat useful, at least as *I* view them.
1) src/Mod/Fem/App/FemSetObject.h includes itself. Thats a first
2)Most places of the code dont care about Gui/Selection.h `SelectionSingleton`, they just need `SelectionObject`s . That is inline with my flamey belief that singletons should die a painful death.
3) <Mod/Part/App/Attacher.h> Looks like it belongs in Base? 
4)Tree.h is bloated, and needs to be chopped up
5)From the doxygen graphs, it looks like Base/Writer+Reader really should be split , zipios++ ends up included in random files without any use.

**Comment by chennes** [comment]
Interesting, I'm not getting a measurable speedup on Windows. I'll try to take a closer look tomorrow, and maybe throw some more hardware at it to get a bigger sample.

**Comment by vosk** [comment]
Is it really worth the trouble? :smile:

**Comment by chennes** [comment]
Probably not, but it might be interesting nonetheless. It's not a top priority, but if I have time I'll still take a look.

**Comment by wwmayer** [comment]
It's best to postpone it until after the release of v0.19.

**Comment by berndhahnebach** [comment]
Following a link to the branch on the CI-repository:

https://gitlab.com/berndhahnebach/FreeCAD/-/commits/PR_5060

The CI-status is available on the latest commit of the branch.
If there is no status available the PR should be rebased on latest master.
Check pipeline branches to see if your PR has been run by the CI.

https://gitlab.com/berndhahnebach/FreeCAD/-/pipelines?scope=branches

**Comment by donovaly** [comment]
It is time to merge this PR. To be able to do this, can you please rebase it?

**Comment by donovaly** [comment]
since @chennes already approved this PR -> assigning him to do the merge

**Comment by chennes** [comment]
Is it still worth doing? In my last test there was no measurable speedup.

**Comment by donovaly** [comment]
> Is it still worth doing? In my last test there was no measurable speedup.

@vosk ?

I would say if there is still speedup for @vosk and no speed loss possible for others it is worth it.

File: src/App/ComplexGeoData.h [hunk 1]
```diff
@@ -23,18 +23,24 @@
...
 #ifdef __GNUC__
 # include <stdint.h>
 #endif
 
+namespace Base
+{
+class Placement;
+class Rotation;
+}
+
```
Comment (vincele) [inline src/App/ComplexGeoData.h:43]: extra blank line

**Review by vincele — COMMENTED** [review]
You may want to have a look at LKML, there's an enormous PR for a massive header rework that yields pretty good build speedup.

Even if this is not speeding the build, it can serve as a good starting point.

I think header untangling is useful work.

But I'm seeing strange diff in GH diff UI, is the patch mangling CRLFs or something like that ? Any edits on windows vs others on linux ?

**Comment by chennes** [comment]
@donovaly Does your recent work on the headers make this PR obsolete?

**Comment by donovaly** [comment]
Thanks for the pointer. This PR is indeed what I almost do. Since I already did a lot, I will review the changes made by this PR and commit them.

**Comment by donovaly** [comment]
I had a look and the PR is inconsistent. There are additions of includes without any comment or obvious reason why this will speedup the compilation.

In the meantime I already removed unused includes and will do so the next days for App too.

**Comment by wwmayer** [comment]
There are some further things to improve:
* in some cases it can be avoided to include <iostream> within a header and replaced with <iosfwd>
* Python.h is included in a lot of header files and when you look at the file you will see that it includes almost another 80 header files which considerably slows down the build process. In many cases the include of Python.h can be replaced with this forward declaration: `typedef struct _object PyObject;`
Then in cases where currently <CXX/Objects.hxx> is included (that also includes Python.h) to use Py::Object a small wrapper for PyObject can be implemented that behaves like a Py::Object.
* For example Console.h contains several classes and thus includes stream and Python stuff. If it's split up into two header & source files then it can be reduced to include the stream classes and Python stuff.

**Comment by donovaly** [comment]
> There are some further things to improve:

Do you plan some work in this regards or should I do something?

**Comment by wwmayer** [comment]
> Do you plan some work in this regards or should I do something?

In a local branch I did something in this regard a few years ago. So, I will have to rework it and merge it to master.

**Comment by donovaly** [comment]
This PR can be closed. A lot has been done recently to reduce the compile time. The PR here cannot be merged and most of it is already in.

**Pull request closed without merging**